# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige werkcollege de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import pandas as pd
import sqlite3
import numpy as np
import warnings
warnings.simplefilter('ignore')

Zet de volgende bestanden in een makkelijk te vinden map, als je dat nog niet gedaan hebt:
- go_sales.sqlite
- go_crm.sqlite
- go_staff.sqlite

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige werkcollege de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
con = sqlite3.connect("go_sales.sqlite")

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [3]:
tabelen = ("product", "product_type", "product_line", "sales_staff", "sales_branch", "retailer_site", "country", "order_header", "order_details", "returned_item", "return_reason")

sqlite_master = {}

for name in tabelen :
   sqlite_master[name] = pd.read_sql(f"SELECT * FROM {name}", con)



Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt.<br> <u>Let op:</u> lees eventueel de informatie uit het vorige werkcollege nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [4]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
pd.read_sql(sql_query, con)
#Op de puntjes hoort de connectie naar go_sales óf go_staff óf go_crm te staan.

type           name       tbl_name  rootpage  \
0   table        country        country       480   
1   table  order_details  order_details        11   
2   table   order_header   order_header        24   
3   table   order_method   order_method        27   
4   table        product        product        28   
5   table   product_line   product_line        17   
6   table   product_type   product_type        10   
7   table  retailer_site  retailer_site        18   
8   table  return_reason  return_reason        30   
9   table  returned_item  returned_item        31   
10  table   sales_branch   sales_branch         2   
11  table    sales_staff    sales_staff         3   

                                                  sql  
0   CREATE TABLE "country" (\n\t"COUNTRY_CODE"\tIN...  
1   CREATE TABLE "order_details" (\n\t"ORDER_DETAI...  
2   CREATE TABLE "order_header" (\n\t"ORDER_NUMBER...  
3   CREATE TABLE "order_method" (\n\t"ORDER_METHOD...  
4   CREATE TABLE "product" (\n\t"PRODUCT_NUMBER"\t...  
5   CREATE TABLE "product_line" (\n\t"PRODUCT_LINE...  
6   CREATE TABLE "product_type" (\n\t"PRODUCT_TYPE...  
7   CREATE TABLE "retailer_site" (\n\t"RETAILER_SI...  
8   CREATE TABLE "return_reason" (\n\t"RETURN_REAS...  
9   CREATE TABLE "returned_item" (\n\t"RETURN_CODE...  
10  CREATE TABLE "sales_branch" (\n\t"SALES_BRANCH...  
11  CREATE TABLE "sales_staff" (\n\t"SALES_STAFF_C...

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
product = sqlite_master["product"]
hogerDan50 = product["PRODUCTION_COST"] > 50
lagerDan100 = product["PRODUCTION_COST"] < 100
product.loc[(hogerDan50 & lagerDan100), :]

PRODUCT_NUMBER INTRODUCTION_DATE  PRODUCT_TYPE_CODE  PRODUCTION_COST  \
5                 6        1997-03-05                  1            85.11   
7                 8        1997-03-05                  1            75.00   
16               17        1997-03-05                  3            60.00   
17               18        1995-02-15                  3            86.00   
19               20        1997-03-05                  3            78.55   
22               23        1997-03-05                  3            65.33   
23               24        1997-03-05                  4            62.50   
45               46        2005-01-10                  7            52.86   
47               48        2005-01-10                  7            55.00   
57               58        2005-01-10                  9            57.13   
68               69        1995-02-15                 11            73.33   
69               70        1997-03-05                 11            80.00   
70               71        1997-03-05                 11            60.00   
71               72        2000-10-26                 11            72.50   
74               75        2000-10-26                 12            80.00   
77               78        1995-02-15                 13            79.19   
78               79        1995-02-15                 13            92.58   
79               80        1997-03-05                 13            94.12   
82               83        1997-03-05                 14            56.25   
83               84        1997-03-05                 14            78.55   
109             110        2003-12-10                 20            60.00   
110             111        2003-12-15                 20            81.80   
113             114        2003-12-15                 21            80.00   

     MARGIN    PRODUCT_IMAGE LANGUAGE                     PRODUCT_NAME  \
5      0.28    P06CE1CG1.jpg       EN        TrailChef Deluxe Cook Set   
7      0.41    P08CE1CG1.jpg       EN           TrailChef Double Flame   
16     0.28    P17CE1SB3.jpg       EN                  Hibernator Lite   
17     0.33    P17CE1SB3.jpg       EN                       Hibernator   
19     0.28    P20CE1SB3.jpg       EN  Hibernator Self - Inflating Mat   
22     0.33    P23CE1SB3.jpg       EN              Hibernator Camp Cot   
23     0.17    P24CE1PK4.jpg       EN     Canyon Mule Climber Backpack   
45     0.28    P46ME2SY7.jpg       EN          Granite Climbing Helmet   
47     0.50    P47ME2SY7.jpg       EN            Husky Harness Extreme   
57     0.28    P58ME2TL9.jpg       EN                   Granite Hammer   
68     0.33   P69PA3EW11.jpg       EN                        Polar Ice   
69     0.33   P70PA3EW11.jpg       EN                     Polar Sports   
70     0.33   P71PA3EW11.jpg       EN                       Polar Wave   
71     0.50   P72PA3EW11.jpg       EN                    Polar Extreme   
74     0.28   P75PA3KV12.jpg       EN                     Edge Extreme   
77     0.20   P78PA3BN13.jpg       EN                        Seeker 35   
78     0.25   P78PA3BN13.jpg       EN                        Seeker 50   
79     0.40   P80PA3BN13.jpg       EN                   Seeker Extreme   
82     0.37   P83PA3NV14.jpg       EN                   Glacier Deluxe   
83     0.28   P84PA3NV14.jpg       EN                      Glacier GPS   
109    0.50  P110GE5PT20.jpg       EN                Blue Steel Putter   
110    0.55  P111GE5PT20.jpg       EN            Blue Steel Max Putter   
113    0.60  P114GE5GA21.jpg       EN              Course Pro Golf Bag   

                                           DESCRIPTION  
5    Cascade set features 1, 2, and 3 liter pots wi...  
7    The TrailChef Double Flame is a camp stove whi...  
16   The Hibernator Lite is the perfect summer slee...  
17   The Hibernator is a three-season sleeping bag....  
19   Mattress is made of extra-thick foam. Mesh stu...  
22   Aluminum frame camp cot that is 

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [ ]:
margeLagerDan20 = product["MARGIN"] < 0.20
margeHogerDan60 = product["MARGIN"] > 0.60
product.loc[(margeHogerDan60 | margeLagerDan20), :]

PRODUCT_NUMBER INTRODUCTION_DATE  PRODUCT_TYPE_CODE  PRODUCTION_COST  \
20              21        1995-02-15                  3            29.17   
23              24        1997-03-05                  4            62.50   
85              86        1995-02-15                 15             2.00   
86              87        1995-02-15                 15             2.00   
87              88        1995-02-15                 15             2.00   
88              89        1995-02-15                 15             2.50   
89              90        1995-02-15                 15             2.50   

    MARGIN   PRODUCT_IMAGE LANGUAGE                  PRODUCT_NAME  \
20    0.17   P21CE1SB3.jpg       EN                Hibernator Pad   
23    0.17   P24CE1PK4.jpg       EN  Canyon Mule Climber Backpack   
85    0.67  P86OP4IR15.jpg       EN             BugShield Natural   
86    0.67  P86OP4IR15.jpg       EN               BugShield Spray   
87    0.70  P86OP4IR15.jpg       EN         BugShield Lotion Lite   
88    0.63  P86OP4IR15.jpg       EN              BugShield Lotion   
89    0.63  P86OP4IR15.jpg       EN             BugShield Extreme   

                                          DESCRIPTION  
20  Warms up quickly for snow camping comfort. The...  
23  This pack is perfect for day trips and short h...  
85  Contains no chemicals. Repels mosquitoes, blac...  
86  Repels black flies, ticks, mosquitoes, chigger...  
87  Repels black flies, ticks, mosquitoes, chigger...  
88  Repels black flies, ticks, mosquitoes, chigger...  
89  Highest amount of the active ingredient, 38%, ...

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [6]:
country = sqlite_master["country"]
country.loc[(country["CURRENCY_NAME"] == "francs" ), :]

COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME
0              1       France       EN        francs
7              8  Switzerland       EN        francs
15            17      Belgium       EN        francs

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [ ]:
product.loc[(product["MARGIN"] > .5), :].drop_duplicates("INTRODUCTION_DATE")

PRODUCT_NUMBER INTRODUCTION_DATE  PRODUCT_TYPE_CODE  PRODUCTION_COST  \
66               67        2000-10-26                 10           116.19   
85               86        1995-02-15                 15             2.00   
101             102        2003-12-10                 18           380.95   
103             104        2003-12-18                 18           333.32   
104             105        2003-12-27                 19           555.90   
106             107        2004-01-13                 19           509.07   
110             111        2003-12-15                 20            81.80   

     MARGIN    PRODUCT_IMAGE LANGUAGE                       PRODUCT_NAME  \
66     0.51   P67PA3WT10.jpg       EN               Mountain Man Extreme   
85     0.67   P86OP4IR15.jpg       EN                  BugShield Natural   
101    0.51  P102GE5IR18.jpg       EN           Hailstorm Titanium Irons   
103    0.51  P102GE5IR18.jpg       EN      Lady Hailstorm Titanium Irons   
104    0.55  P105GE5WD19.jpg       EN       Hailstorm Titanium Woods Set   
106    0.55  P105GE5WD19.jpg       EN  Lady Hailstorm Titanium Woods Set   
110    0.55  P111GE5PT20.jpg       EN              Blue Steel Max Putter   

                                           DESCRIPTION  
66   Solid titanium case, crown, bezel.  Combinatio...  
85   Contains no chemicals. Repels mosquitoes, blac...  
101  Made entirely of pure titanium. The ultimate i...  
103  Made entirely of pure titanium. A wider sole a...  
104  Crafted of 6-4 titanium which is 40% lighter a...  
106  Crafted of 6-4 titanium which is 40% lighter a...  
110  Putter head is composed from a single piece of...

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [7]:
sales_branch = sqlite_master["sales_branch"]
sales_branch.loc[(sales_branch["ADDRESS2"].notna() & sales_branch["REGION"].notna()), ("ADDRESS1", "CITY")]

ADDRESS1              CITY
2                Singelgravenplein 4         Amsterdam
13  Prol. Paseo de la Reforma No. 51  Distrito Federal
14               202-2-3 Hyakunincho             Tokyo
15                     543-225 Asahi        Osaka City
16                  2315 Queen's Ave         Melbourne
18             Avenida Paulista, 333         São Paulo
24                    3 Albany Court        Birmingham

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [ ]:
country.loc[((country["CURRENCY_NAME"] == "dollars") | (country["CURRENCY_NAME"] == "new dollar")),  :].sort_values("COUNTRY")

COUNTRY_CODE        COUNTRY LANGUAGE CURRENCY_NAME
14            16      Australia       EN       dollars
3              4         Canada       EN       dollars
11            12         Taiwan       EN    new dollar
2              3  United States       EN       dollars

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [8]:
retailer_site = sqlite_master["retailer_site"]
retailer_site.loc[((retailer_site["POSTAL_ZONE"].str[0] == "D") & (retailer_site["ADDRESS2"].notna())), ("ADDRESS1", "ADDRESS2", "CITY")]

ADDRESS1       ADDRESS2       CITY
31  Röntgenstraße 90  3. Tür rechts  Frankfurt
59   Grubesallee 141       4. Stock    Hamburg

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [14]:
returned_item = sqlite_master["returned_item"]
returned_item.count()["RETURN_CODE"]

np.int64(619)

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [ ]:
sales_branch["REGION"].nunique()

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [ ]:
product["MARGIN"].agg([min, max, np.mean]).round(2)

min     0.17
max     0.70
mean    0.40
Name: MARGIN, dtype: float64

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [ ]:
retailer_site["ADDRESS2"].isna().sum()

np.int64(286)

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [9]:
order_details = sqlite_master["order_details"]
order_details.loc[(order_details["UNIT_SALE_PRICE"] < order_details["UNIT_PRICE"]), "UNIT_COST"].agg(np.mean)

np.float64(127.34087927598803)

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [10]:
sales_staff = sqlite_master["sales_staff"]
sales_staff.groupby("POSITION_EN", as_index=False)["SALES_STAFF_CODE"].count()

POSITION_EN  SALES_STAFF_CODE
0                Branch Manager                19
1              District Manager                 4
2               General Manager                 1
3  Level 1 Sales Representative                13
4  Level 2 Sales Representative                39
5  Level 3 Sales Representative                24
6              Regional Manager                 2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [ ]:
test = sales_staff.groupby("WORK_PHONE")["WORK_PHONE"].count() > 4
test.loc[test]

WORK_PHONE
+(44) 121 3505267      True
+(61) 03 2982  4242    True
+31 (0)20 692 93 94    True
+33 1 68 94 52 20      True
1 (206) 292-0012       True
1 (305) 557-4810       True
1 (310) 281-5722       True
1 (403) 232-5986       True
1 (416) 493-5595       True
1 (617) 268-6754       True
Name: WORK_PHONE, dtype: bool

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [ ]:
NLklant = pd.merge(country, retailer_site, on="COUNTRY_CODE")
NLklant.loc[(NLklant["COUNTRY_CODE"] == 7), :]

COUNTRY_CODE      COUNTRY LANGUAGE CURRENCY_NAME  RETAILER_SITE_CODE  \
215             7  Netherlands       EN      guilders                 272   
216             7  Netherlands       EN      guilders                 273   
217             7  Netherlands       EN      guilders                 274   
218             7  Netherlands       EN      guilders                 275   
219             7  Netherlands       EN      guilders                 276   
220             7  Netherlands       EN      guilders                 277   
221             7  Netherlands       EN      guilders                 278   
222             7  Netherlands       EN      guilders                 279   
223             7  Netherlands       EN      guilders                 280   
224             7  Netherlands       EN      guilders                 281   
225             7  Netherlands       EN      guilders                 282   
226             7  Netherlands       EN      guilders                 283   
227             7  Netherlands       EN      guilders                 284   
228             7  Netherlands       EN      guilders                 285   
229             7  Netherlands       EN      guilders                 286   
230             7  Netherlands       EN      guilders                 287   
231             7  Netherlands       EN      guilders                 288   
232             7  Netherlands       EN      guilders                 289   
233             7  Netherlands       EN      guilders                 290   
234             7  Netherlands       EN      guilders                 291   

     RETAILER_CODE                   ADDRESS1 ADDRESS2          CITY REGION  \
215            141                   Beets 36      NaN         Beets    NaN   
216            141               Perestraat 1      NaN      Deventer    NaN   
217            141  Laan van Meerdervoort 966      NaN      Den Haag    NaN   
218            141             Kalfjeslaan 47      NaN    Amstelveen    NaN   
219            141               Startbaan 11      NaN    Amstelveen    NaN   
220            142      Noorderspoorsingel 28      NaN     Groningen    NaN   
221            143              Hogehilweg 13      NaN     Amsterdam    NaN   
222            143             Spoorstraat 43      NaN    Varsseveld    NaN   
223            144             Westhavenweg 8      NaN     Amsterdam    NaN   
224            144              Westzeedijk 4      NaN        Muiden    NaN   
225            144            Gooiergracht 91      NaN         Laren    NaN   
226            144            Hoofdstraat 150      NaN   Pieterburen    NaN   
227            145                Olde Diek 2      NaN         Laren    NaN   
228            145     Amsterdamsestraatweg 8      NaN      Maarssen    NaN   
229            145        Krayenhoffstraat 11      NaN     Amsterdam    NaN   
230            145         Overschiestraat 19      NaN     Amsterdam    NaN   
231            145              Havenbuurt 12      NaN        Marken    NaN   
232            146               Duivendaal 1      NaN    Wageningen    NaN   
233            146              Voorstraat 37      NaN  Nieuweschans    NaN   
234            146         Stadhouderslaan 18      NaN     Rotterdam    NaN   

    POSTAL_ZONE  ACTIVE_INDICATOR  
215     1475 JC                 1  
216     7412 BV                 1  
217     2564 AV                 1  
218     1181 BN                 1  
219     1185 XP                 1  
220     9716 JE                 1  
221     1000 BA                 1  
222     7051 CH                 1  
223     1014 BD                 1  
224     1398 BB                 1  
225     1251 VC                 1  
226     9968 AH                 1  
227     7245 VS                 1  
228     3604 AA                 1  
229     1018 DE                 1  
230     1062 XC                 1  
231     1156 AL                 1  
232     6701 AP                 1  
233     9693 EE                 1  
234    

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [11]:
product_type = sqlite_master["product_type"]
Eyewear = pd.merge(product, product_type, on= "PRODUCT_TYPE_CODE")
Eyewear.loc[(Eyewear["PRODUCT_TYPE_EN"] == "Eyewear"), :]

PRODUCT_NUMBER INTRODUCTION_DATE  PRODUCT_TYPE_CODE  PRODUCTION_COST  \
67              68        1995-02-15                 11            46.15   
68              69        1995-02-15                 11            73.33   
69              70        1997-03-05                 11            80.00   
70              71        1997-03-05                 11            60.00   
71              72        2000-10-26                 11            72.50   

    MARGIN   PRODUCT_IMAGE LANGUAGE   PRODUCT_NAME  \
67    0.23  P68PA3EW11.jpg       EN      Polar Sun   
68    0.33  P69PA3EW11.jpg       EN      Polar Ice   
69    0.33  P70PA3EW11.jpg       EN   Polar Sports   
70    0.33  P71PA3EW11.jpg       EN     Polar Wave   
71    0.50  P72PA3EW11.jpg       EN  Polar Extreme   

                                          DESCRIPTION  PRODUCT_LINE_CODE  \
67  The Polar Sun sunglasses have 3 mm thick lense...                  3   
68  Polar Ice sunglasses have polarized lenses to ...                  3   
69  Six base classic nylon composite frame sunglas...                  3   
70  Acrylex lenses in the Polar Wave sunglasses ab...                  3   
71  Ideal for skiing and winter mountaineering. Fr...                  3   

   PRODUCT_TYPE_EN  
67         Eyewear  
68         Eyewear  
69         Eyewear  
70         Eyewear  
71         Eyewear

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [12]:
order_header = sqlite_master["order_header"]
merge1 = pd.merge(order_header, retailer_site)
merge2 = pd.merge(merge1, sales_staff)
merge2.loc[((merge2["POSITION_EN"] == "Branch Manager")), ("ADDRESS1", "FIRST_NAME", "LAST_NAME")].drop_duplicates("ADDRESS1")

ADDRESS1 FIRST_NAME LAST_NAME
3185  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 

In [20]:
merge1 = pd.merge(order_header, sales_staff, how="outer")
merge1.loc[(merge1["POSITION_EN"].str.contains("Manager")), ("POSITION_EN", "ORDER_DATE")].drop_duplicates()

POSITION_EN  ORDER_DATE
0       Branch Manager         NaN
320   Regional Manager         NaN
648   District Manager         NaN
3748    Branch Manager  2024-09-09
3749    Branch Manager  2024-03-13
3750    Branch Manager  2024-01-09
4186   General Manager         NaN

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [23]:
merge1 = pd.merge(product, product_type)
merge2 = pd.merge(merge1, order_details)
merge2.loc[(merge2["QUANTITY"] > 750), ("PRODUCT_NAME", "PRODUCT_TYPE_EN")].drop_duplicates()

PRODUCT_NAME    PRODUCT_TYPE_EN
5427                Star Peg              Tents
11452              Firefly 2           Lanterns
28608        BugShield Spray  Insect Repellents
28892  BugShield Lotion Lite  Insect Repellents
29430       BugShield Lotion  Insect Repellents
29442      BugShield Extreme  Insect Repellents
30741         Sun Shelter 15          Sunscreen
31164         Sun Shelter 30          Sunscreen
31775             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [25]:
merge1 = pd.merge(product, order_details)
merge1.loc[((merge1["UNIT_PRICE"] - merge1["UNIT_SALE_PRICE"]) / merge1["UNIT_PRICE"] > 0.4), ("PRODUCT_NAME")].drop_duplicates()

27830               BugShield Natural
28283                 BugShield Spray
28694           BugShield Lotion Lite
29127                BugShield Lotion
29464               BugShield Extreme
30562                  Sun Shelter 15
30929                  Sun Shelter 30
34441    Hailstorm Titanium Woods Set
Name: PRODUCT_NAME, dtype: str

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [15]:
return_reason = sqlite_master["return_reason"]
merge1 = pd.merge(returned_item, return_reason)
merge2 = pd.merge(merge1, order_details)
merge2.loc[((merge2["RETURN_QUANTITY"] / merge2["QUANTITY"]) > 0.9), "RETURN_DESCRIPTION_EN"].drop_duplicates()

0    Unsatisfactory product
2     Wrong product shipped
3     Wrong product ordered
Name: RETURN_DESCRIPTION_EN, dtype: str

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [16]:
merge1 = pd.merge(product_type, product)
merge1 = merge1["PRODUCT_TYPE_EN"].value_counts()
merge1

PRODUCT_TYPE_EN
Lanterns                12
Cooking Gear            10
Sleeping Bags            7
Climbing Accessories     7
Tents                    6
Packs                    6
Tools                    6
Watches                  5
Eyewear                  5
Knives                   5
Insect Repellents        5
Sunscreen                5
First Aid                5
Rope                     4
Safety                   4
Binoculars               4
Navigation               4
Irons                    4
Woods                    4
Golf Accessories         4
Putters                  3
Name: count, dtype: int64

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [17]:
merge1 = pd.merge(country, sales_branch)
merge1 = merge1["COUNTRY"].value_counts()
merge1

COUNTRY
United States     4
France            2
Germany           2
Canada            2
United Kingdom    2
Japan             2
Austria           1
Italy             1
Netherlands       1
Switzerland       1
Sweden            1
Taiwan            1
Korea             1
China             1
Australia         1
Belgium           1
Spain             1
Mexico            1
Brazil            1
Finland           1
Name: count, dtype: int64

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (6 kolommen, 10 rijen)

In [52]:
merge1 = pd.merge(product_type, product)
merge2 = pd.merge(merge1, order_details)
merge2.loc[merge2["PRODUCT_TYPE_EN"] == "Cooking Gear"].groupby(["PRODUCT_NAME", "PRODUCT_TYPE_EN", "PRODUCT_NUMBER", "DESCRIPTION"], as_index=False)["UNIT_SALE_PRICE"].agg(["mean", "count"]).sort_values("count", ascending=False)


PRODUCT_NAME PRODUCT_TYPE_EN  PRODUCT_NUMBER  \
9        TrailChef Water Bag    Cooking Gear               1   
5           TrailChef Kettle    Cooking Gear               9   
6      TrailChef Kitchen Kit    Cooking Gear               3   
4     TrailChef Double Flame    Cooking Gear               8   
8         TrailChef Utensils    Cooking Gear              10   
0          TrailChef Canteen    Cooking Gear               2   
2              TrailChef Cup    Cooking Gear               4   
1         TrailChef Cook Set    Cooking Gear               5   
7     TrailChef Single Flame    Cooking Gear               7   
3  TrailChef Deluxe Cook Set    Cooking Gear               6   

                                         DESCRIPTION        mean  count  
9  Lightweight, collapsible bag to carry liquids ...    6.032506    419  
5  This old-fashioned style kettle is made of dur...   12.201235    413  
6  Zippered nylon pouch contains cutlery for two,...   22.345975    400  
4  The TrailChef Double Flame is a camp stove whi...  132.672882    399  
8  Spoon, fork and knife set made of a light yet ...   17.294151    383  
0  Aluminum canteen. Rugged fleece-lined cover wi...   11.979418    361  
2            Tin cup. Holds 0.4 liters. Weight: 60 g    6.976490    359  
1  All you will ever need on the trail.  Pot grip...   51.687926    352  
7  A single burner stove with integral tank.  It ...   63.827622    307  
3  Cascade set features 1, 2, and 3 liter pots wi...  121.217638    254

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [90]:
merge1 = pd.merge(country, sales_branch)
merge1.assign(klanten=merge1.groupby("COUNTRY")["CITY"].transform("nunique"))[["COUNTRY", "CITY", "klanten"]].rename(columns={"CITY": "verkoper"})

COUNTRY          verkoper  klanten
0           France             Paris        2
1           France              Lyon        2
2          Germany           Hamburg        2
3          Germany           München        2
4    United States            Boston        4
5    United States           Seattle        4
6    United States       Los Angeles        4
7    United States             Miami        4
8           Canada           Calgary        2
9           Canada           Toronto        2
10         Austria              Wien        1
11           Italy            Milano        1
12     Netherlands         Amsterdam        1
13     Switzerland            Zürich        1
14  United Kingdom            London        2
15  United Kingdom        Birmingham        2
16          Sweden             Kista        1
17           Japan             Tokyo        2
18           Japan        Osaka City        2
19          Taiwan         San Chung        1
20           Korea             Seoul        1
21           China          Shanghai        1
22       Australia         Melbourne        1
23         Belgium          Heverlee        1
24           Spain            Bilbao        1
25          Mexico  Distrito Federal        1
26          Brazil         São Paulo        1
27         Finland            Kuopio        1

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [107]:
merge1 = pd.merge(sales_staff, order_header, how="left")
merge1.loc[~merge1["SALES_STAFF_CODE"].isin(order_header["SALES_STAFF_CODE"]) , ["FIRST_NAME", "LAST_NAME"]].drop_duplicates()

FIRST_NAME      LAST_NAME
0         Denis           Pagé
320        Else         Mörike
321       Frank          Fuchs
647       Fritz         Hirsch
648        Jörg          Kunze
728       Maria      Iacobucci
946        Kick        Kalkman
1309      Karin      Bergström
1395      Sally          White
1421      Frank        Bretton
1422    Georges  Saint-Germain
1719       Alex      Rodriguez
2088   Samantha          Floyd
2392      Julie          Olsen
2530      Corey         Wright
2873    Estelle        Johnson
2874       Greg        Belding
2999      Julie        Béchard
3157      Lucía          Reyna
3401     Kazumi        Uragome
3559      Alice         Walter
4185     Walter         Taylor
4186   Penelope        Tamrine
4349      Polly         Linton
4590      Gracy        Gellens

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [59]:
pd.DataFrame({
    "overzicht": [
        (product["MARGIN"] < product["MARGIN"].mean()).sum(),
        product.loc[product["MARGIN"] < product["MARGIN"].mean(), "MARGIN"].mean()
    ]
})


overzicht
0  59.000000
1   0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [114]:
merge1 = pd.merge(order_details, returned_item, how='left')
merge2 = pd.merge(merge1, product, how='left')
merge2.loc[(~merge2["ORDER_DETAIL_CODE"].isin(returned_item["ORDER_DETAIL_CODE"])) & (merge2["UNIT_PRICE"] > 500), ["PRODUCT_NAME"]].drop_duplicates().reset_index(drop=True)

PRODUCT_NAME
0                           Star Dome
1                        Star Gazer 2
2                        Star Gazer 3
3                        Star Gazer 6
4               Hailstorm Steel Irons
5            Hailstorm Titanium Irons
6          Lady Hailstorm Steel Irons
7       Lady Hailstorm Titanium Irons
8        Hailstorm Titanium Woods Set
9           Hailstorm Steel Woods Set
10  Lady Hailstorm Titanium Woods Set
11     Lady Hailstorm Steel Woods Set
12                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [129]:
sales_staff.assign(Manager=np.where(sales_staff["POSITION_EN"].str.contains("Manager"), "Ja", "Nee"))[["LAST_NAME", "Manager"]]

LAST_NAME Manager
0         Pagé      Ja
1       Michel     Nee
2     Clermont     Nee
3       Jauvin     Nee
4    Wiesinger     Nee
..         ...     ...
97    Laermans     Nee
98     De Crée     Nee
99     Lattrez     Nee
100  Seefelder     Nee
101     Grüner     Nee

[102 rows x 2 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [115]:
from datetime import date
date.today().year

2026

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [134]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 30 jaar in dienst) of een ‘lang in dienst’ (groter dan gelijk aan 30 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [140]:
sales_staff.assign(Dienst=np.where((pd.to_datetime(date.today()) - pd.to_datetime(sales_staff["DATE_HIRED"])).dt.days >= 30*365.25 ,"lang in dienst", "kort in dienst"))[["FIRST_NAME", "Dienst"]]

FIRST_NAME          Dienst
0        Denis  kort in dienst
1    Élizabeth  lang in dienst
2        Émile  kort in dienst
3      Étienne  kort in dienst
4      Elsbeth  kort in dienst
..         ...             ...
97       Giele  kort in dienst
98    François  kort in dienst
99      Yvette  kort in dienst
100      Willi  kort in dienst
101     Sabine  kort in dienst

[102 rows x 2 columns]